src/pybind/mgr/dashboard/controllers/auth_classes.py

```diff
@@ -0,0 +1,88 @@
+import abc
+import base64
+import json
+from enum import Enum
+from typing import Dict
+
+
+class AuthType(str, Enum):
+    LOCAL = ''
+    SAML2 = 'saml2'
+    OAUTH2 = 'oauth2'
+
+
+class BaseAuth(abc.ABC):
+    LOGIN_URL: str
+    LOGOUT_URL: str
+    sso: bool
+
+    @classmethod
+    def from_protocol(cls, protocol: AuthType) -> "BaseAuth":
+        auth_map = {
+            AuthType.LOCAL: Local,
+            AuthType.SAML2: Saml2,
+            AuthType.OAUTH2: OAuth2
+        }
+        return auth_map[protocol]
+
+
+class Local(BaseAuth):
+    LOGIN_URL = '#/login'
+    LOGOUT_URL = '#/login'
+    sso = False
+
+
+class SSOAuth(BaseAuth):
+    sso = True
+
+    def to_dict(self) -> Dict[str, str]:
+        return {}
+
+    @classmethod
+    def from_dict(cls, s_dict) -> 'SSOAuth':
+        # pylint: disable=unused-argument
+        return cls()
+
+
+class Saml2(SSOAuth):
+    LOGIN_URL = 'auth/saml2/login'
+    LOGOUT_URL = 'auth/saml2/slo'
+    sso = True
+
+    def __init__(self, onelogin_settings):
+        self.onelogin_settings = onelogin_settings
+
+    def get_username_attribute(self):
+        return self.onelogin_settings['sp']['attributeConsumingService']['requestedAttributes'][0][
+            'name']
+
+    def to_dict(self) -> Dict[str, str]:
+        return {
+            'onelogin_settings': self.onelogin_settings
+        }
+
+    @classmethod
+    def from_dict(cls, s_dict):
+        return Saml2(s_dict['onelogin_settings'])
+
+
+class OAuth2(SSOAuth):
+    LOGIN_URL = 'auth/oauth2/login'
+    LOGOUT_URL = 'auth/oauth2/logout'
+    sso = True
+
+    def __init__(self):
+        pass
+
+    def to_dict(self) -> Dict[str, str]:
+        return {}
+
+    @classmethod
+    def from_dict(cls, s_dict):
+        return super().from_dict(s_dict)
```

Comment:
Same thing here: if we're expecting some dict with specific fields, then we should declare it.